LB3210.org

```diff
@@ -977,6 +982,208 @@ Route route = path("users", () ->
+#+BEGIN_SRC restclient
+GET http://localhost/journal/events?since=Wed+May+26+11:59:05+2021+UTC
+Accept: text/event-stream
+#+END_SRC
+  * Web sockets
+#+BEGIN_SRC restclient
+GET http://localhost/journal/events?since=Wed+May+26+11:59:05+2021+UTC
+Connection: Upgrade
+Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==
+Sec-WebSocket-Protocol: chat, superchat
+Sec-WebSocket-Version: 13
+
+HTTP/1.1 101 Switching Protocols
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Accept: HSmrc0sMlYUkAGmm5OPpG2HaGWk=
+Sec-WebSocket-Protocol: chat
+#+END_SRC
+** Multi-dimensional versioning
+*** Semantic versioning in REST
+- Often used for library dependencies and packaged software releases
+- Version number has three parts (/major/, /minor/, /patch/): version =2.0.15=
+  * A new release always must have a new version
+  * If a release has no new functionality (only bugfixes), increase the /patch/
+  * If a release has new functionality that doesn't break API promises, increase the /minor/
+  * If a release has new breaking functionality, increase the /major/
+- How does this relate to REST?
+*** Semantic versioning in REST (cont.)
+- How does this relate to REST?
+  * It doesn't!
+  * REST is a call to a remote system
+    + Could be deploying new versions multiple times per day
+  * The whole point is the client /doesn't/ want (or need) to see those
+
+- OK, what do we do instead?
+  * Version across all HTTP dimensions
+*** Versioning in body structure
+- Many extensions fit fine into existing body structure
+  * Adding of fields
+  * Adding of values to enumerations or code lists
+- If DDD has done its work, terminology should mostly hold
+*** Versioning in content type
+- If a breaking change is needed
+- It might be limited to only one content type
+- Client requests old version:
+#+BEGIN_SRC restclient
+GET http://localhost/myresource
+Accept: application/vnd.example.myresource.v1+json
+#+END_SRC
+- Client requests new version:
+#+BEGIN_SRC restclient
+GET http://localhost/myresource
+Accept: application/vnd.example.myresource.v2+json
+#+END_SRC
+*** Versioning in query parameters
+- Don't do this
+  * Query parameters affect /which/ and /what/ resource(s) are returned, not /how/
+- The meaning of query parameters may themselves be versioned
+*** Versioning in path
+#+BEGIN_SRC restclient
+GET http://localhost/service/versions/1/myresource
+#+END_SRC
+- Often used as first choice
+- Should be your last resort:
+  * Your path is the name of your resource
+  * Your DDD workshop (probably) didn't event storm about "versions"
+  * Your system (probably) doesn't have 2 complete implementations
+  * This does often not reflect reality
+*** Versioning using custom headers
+- Client sends a custom header of the API version they've implemented against
+- Server sends a custom header of the API version that's current
+- This does kinda work
+
+- Fairly weak way to work around /actually/ dealing with semantic changes and compatibility
```

Comment:
consider adding references and examples to best-of-breed API's our there, e.g. Github, tradeshift (I guess you know this one pretty well:-), e-conomic, stripe.com.